modules/salt/pages/custom-states.adoc

```diff
@@ -0,0 +1,151 @@
+
+You can also manage revisions, compare the state to others in your organization, and download the ``.sls`` file from this dialog.
+
+
+
+.Procedure: Assigning a Client to a Custom Salt State
+. In the {productname} {webui}, navigate to menu:Configuration[Channels].
+. Click the name of the state you want to assign a client to.
+. Navigate to the menu:Systems[Target Systems] tab.
+. Check the clients you want to assign.
+. Click btn:[Subscribe systems].
+
+
+
+== Example Custom State Files
+
+This section contains some example custom state files.
+Use these as a basis for writing your own custom states.
+
+
+
+.Example: Initial State File
+
+----
+file.managed:
+  - source: salt://example_org/example_channel/etc/motd
+[
+  FIXME: reference other file here somehow
+]
+----
+
+
+
+.Example: Custom State File
+
+----
+install required packages:
+  pkg.installed:
+    - pkgs:
+      - lsof
+    - failhard: True
+update packages:
+  module.run:
+    - name: pkg.upgrade
+    - failhard: True
+check if reboot is needed:
+  salt.set_reboot_needed:
+    - failhard: True
+reboot:
+   salt.reboot_if_needed:
+     - failhard: True
+----
+
+
+
+== Custom State to Trust a GPG Key
+
+By default, operating systems trust only their own GPG keys when they are installed, and do not trust keys provided by third party packages.
+The clients can be successfully bootstrapped without the GPG key being trusted.
+However, you cannot install new client tool packages or update them until the keys are trusted.
+
+Salt clients are set to trust {suse} tools channels GPG keys when they are bootstrapped.
+For all other clients and channels, you need to manually trust third party GPG keys.
+
+If you are bootstrapping Salt clients from the {productname} {webui}, you can use a custom Salt state to trust the GPG key.
+
+
+
+.Procedure: Trusting a GPG Key With a Custom Salt State
+. On the {productname} Server, at the command prompt, check the contents of the [path]``/srv/www/htdocs/pub/`` directory.
+  This directory contains all available public keys.
+  Take a note of the key that applies to the channel assigned to the client you are registering.
+. In the {productname} {webui}, navigate to menu:Configuration[Channels].
+. Click btn:[Create State Channel].
+. In the [guimenu]``Name`` field, type a name for your state.
+  For example, ``GPG Key Trusts``
+. In the [guimenu]``Label`` field, type a label.
+  For example, ``GPG_Key_Trusts``
+. In the [guimenu]``Description`` field, type a short description of the configuration your state performs.
+  For example, ``Trusts GPG Keys for CentOS``.
+. In the [guimenu]``SLS Contents`` field, create a state to retrieve the appropriate key from the {productname} Server and trust it on the client.
+  For example FIXME:
++
+----
+cmd.run:
+  - salt '*' gpg.import_key filename='/srv/www/htdocs/pub/centos8-gpg-pubkey-05B555B38483C65D.key'
+  - salt '*' gpg.trust_key keyid='centos8-gpg-pubkey-05B555B38483C65D.key' trust_level='trusted'
```

Comment:
Using Salt to call Salt seems a bit wird to me... Isn't there something more elegant and straightforward?

Reply:
This was a wild guess on my part. I need someone who knows what they're doing here :stuck_out_tongue:

Reply:
Perhaps @juliogonzalez can help?

Reply:
@juliogonzalez if you can't help, or don't have bandwidth, can you recommend someone else who could?

Reply:
I would need to check, but i fear this would not do what you expect.

Reply:
No, I wasn't really expecting that it would. Who could tell me?

Reply:
Problem: It would be better if we would have a salt module for trusting GPG keys on different OSes, but this does not exist yet. So trusting GPG keys depends on the OS!

```
rpm_trust_gpg_key:
  cmd.run:
    - name: rpm --import https://{{ salt['pillar.get']('mgr_server') }}/pub/third-party-gpg.key
    - unless: rpm -q gpg-pubkey-<key_id>

deb_trust_gpg_key:
  mgrcompat.module_run:
    - name: pkg.add_repo_key
    - path: https://{{ salt['pillar.get']('mgr_server') }}/pub/third-party-gpg.key
```

Reply:
Of cause you can also add the GPG key files to the config channel, use a `file.managed` to deploy them on the minion and use a local path instead of a `https://` URL .